core/controllers/reader.py

```diff
@@ -1371,6 +1371,172 @@ def get(self):
+        'unique_progress_url_id': {
+            'schema': {
+                    'type': 'basestring',
+                }
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'GET': {}
+    }
+
+    @acl_decorators.can_play_exploration
+    def get(self, unique_progress_url_id):
+        """Handles GET requests. Fetches the logged-out learner's progress."""
+
+        logged_out_user_data = (
+            exp_fetchers.get_logged_out_user_progress(unique_progress_url_id))
+
+        if logged_out_user_data is None:
+            raise self.PageNotFoundException()
+
+        exploration = exp_fetchers.get_exploration_by_id(
+            logged_out_user_data.exploration_id, strict=False)
+
+        if logged_out_user_data is not None:
+            synced_exp_user_data = None
+            # If the latest exploration version is ahead of the most recently
+            # interacted exploration version.
+            if (
+                logged_out_user_data.most_recently_reached_checkpoint_exp_version is not None and # pylint: disable=line-too-long
+                logged_out_user_data.most_recently_reached_checkpoint_exp_version < exploration.version # pylint: disable=line-too-long
+            ):
+                synced_exp_user_data = (
+                    exp_services.sync_learner_checkpoint_progress_with_current_exp_version( # pylint: disable=line-too-long
+                        logged_out_user_data.exploration_id, unique_progress_url_id)) # pylint: disable=line-too-long
+            else:
+                synced_exp_user_data = logged_out_user_data
+
+            furthest_reached_checkpoint_exp_version = (
+                synced_exp_user_data.furthest_reached_checkpoint_exp_version)
+            furthest_reached_checkpoint_state_name = (
+                synced_exp_user_data.furthest_reached_checkpoint_state_name)
+            most_recently_reached_checkpoint_exp_version = (
+                synced_exp_user_data
+                    .most_recently_reached_checkpoint_exp_version)
+            most_recently_reached_checkpoint_state_name = (
+                synced_exp_user_data
+                    .most_recently_reached_checkpoint_state_name)
+
+        self.values.update({
+            'exploration_id': (
+                logged_out_user_data.exploration_id),
+            'furthest_reached_checkpoint_exp_version': (
+                furthest_reached_checkpoint_exp_version),
+            'furthest_reached_checkpoint_state_name': (
+                furthest_reached_checkpoint_state_name),
+            'most_recently_reached_checkpoint_exp_version': (
+                most_recently_reached_checkpoint_exp_version),
+            'most_recently_reached_checkpoint_state_name': (
+                most_recently_reached_checkpoint_state_name),
+        })
+
+        self.render_json(self.values)
+
+
+class SaveTransientCheckpointProgressHandler(base.BaseHandler):
+    """Responsible for storing progress of a logged-out learner."""
```

Comment:
Can you be more specific in the docstring that it's for storing checkpoints progress

Reply:
Done.